File: Graph/1514.Path_with_Maximum_Probability/C++/1514.Path_with_Maximum_Probability.cpp
class Solution {
public:
    double maxProbability(int n, vector<vector<int>> & edges, vector<double> & succProb, int start_node, int end_node) 
    {
        //=============================================//
        // Input type:                                 //
        //   - int (n-th nodes)                        //
        //   - vector<vector<int>> ref. (edges vector) //
        //   - vector<double> ref. (succProb vector)   //
        //   - int (start node)                        //
        //   - int (end node)                          //
        // Return type:                                //
        //   - double (result maximun probalility)     //
        //=============================================//
        
        /*Initialize*/
        ///// Length of edges vector /////
        int len_edges = edges.size();

        ///// Record dictionary, hashmap /////
        unordered_map<int, vector<pair<int, double>>> record_map;

        ///// Record heap, priority-queue /////
        priority_queue<pair<double, int>> record_pq;

        ///// Product summary-path vector /////
        vector<double> prod_path_vect(n, 0.0);

        ///// Record value /////
        int record_val;

        ///// Record distance, next distance /////
        double record_dist, next_dist;


        /*Dijkstra's based loop traversal with recorded vectors, dictionary, hashmap*/
        ///// Step 1: Record root-child formations with dictionary, hashmap /////
        for (int edges_idx = 0; (edges_idx < len_edges); (edges_idx++)) //Whole
        {
            (record_map[(edges[edges_idx])[0]]).push_back(make_pair((edges[edges_idx])[1], succProb[edges_idx])); //Keep updating, recording
            (record_map[(edges[edges_idx])[1]]).push_back(make_pair((edges[edges_idx])[0], succProb[edges_idx])); //Keep updating, recording

        } //Whole

        ///// Step 2: Looped-traversal with recorded dictionary, hashmap /////
        record_pq.push(make_pair(1.0, start_node)); //Update, Record

        prod_path_vect[start_node] = 1.0; //Update, Overwrite

        while (! record_pq.empty()) //Cycled-times
        {
            record_val = (record_pq.top()).second; record_dist = (record_pq.top()).first; //Record values, distances, ...

            record_pq.pop(); //Keep updating, popped

            ///// Check if the current node is valid or not /////
            if (prod_path_vect[record_val] != record_dist) { continue; }

            for (pair<int, double> & record_pair: record_map[record_val]) //Next-nodes
            {
                next_dist = (prod_path_vect[record_val] * record_pair.second); //Next distance

                ///// Check if the next distance is larger or not /////
                if (next_dist > prod_path_vect[record_pair.first])
                {
                    record_pq.push(make_pair(next_dist, record_pair.first)); //Keep updating, recording

                    prod_path_vect[record_pair.first] = next_dist; //Keep updating, overwriting
                }
                else { ; }

            } //Next-nodes

        } //Cycled-times

        return prod_path_vect[end_node];
    }
};

File: Graph/1514.Path_with_Maximum_Probability/Python/1514.Path_with_Maximum_Probability.py
class Solution(object):
    def maxProbability(self, n, edges, succProb, start_node, end_node):
        """
        :type n: int
        :type edges: List[List[int]]
        :type succProb: List[float]
        :type start_node: int
        :type end_node: int
        :rtype: float
        """
        #===================================#
        # Dijkstra's based traversal method #
        #===================================#

        ############
        #Initialize
        ##### Length of edges array #####
        len_edges = len(edges)

        ##### Record dictionary/hashmap #####
        record_dict = {}

        ##### Edge rank array #####
        edge_rank_arry = [0.0 for n_idx in range(n)]

        ##### Record heap/priority-queue #####
        record_pq = []


        #######################################################################################
        #Dijkstra's based loop traversal with recorded heap/priority-queue, dictionary/hashmap
        heapify(record_pq) #Update/Transform

        edge_rank_arry[start_node] = 1.0 #Update/Overwrite

        heappush(record_pq, [start_node, (-1.0)]) #Update/Record

        for edges_idx in range(len_edges):

            ##### Check if the current node existed or not #####
            if ((edges[edges_idx])[0] not in record_dict):
                record_dict[(edges[edges_idx])[0]] = [[(edges[edges_idx])[1], succProb[edges_idx]]] #Keep updating/recording
            else:
                (record_dict[(edges[edges_idx])[0]]).append([(edges[edges_idx])[1], succProb[edges_idx]]) #Keep updating/recording
            
            if ((edges[edges_idx])[1] not in record_dict):
                record_dict[(edges[edges_idx])[1]] = [[(edges[edges_idx])[0], succProb[edges_idx]]] #Keep updating/recording
            else:
                (record_dict[(edges[edges_idx])[1]]).append([(edges[edges_idx])[0], succProb[edges_idx]]) #Keep updating/recording

        while (record_pq):
            [record_val, record_scl] = record_pq[0] #Nodes informations

            heappop(record_pq) #Keep updating/popped

            ##### Check if the current node information is valid or not #####
            if (edge_rank_arry[record_val] != (record_scl * (-1))):
                continue

            ##### Check if the current node existed or not #####
            if (record_val in record_dict):
                for [next_val, next_scl] in record_dict[record_val]:
                    next_accu_scl = ((record_scl * (-1)) * next_scl)

                    ##### Check if the next node matched conditions or not #####
                    if (next_accu_scl > edge_rank_arry[next_val]):
                        edge_rank_arry[next_val] = next_accu_scl #Keep updating/overwriting

                        heappush(record_pq, [next_val, (next_accu_scl * (-1))]) #Keep updating/recording
                    else:
                        pass

        return edge_rank_arry[end_node]
